src/Features/Core/Portable/AddParameter/AbstractAddParameterCodeFixProvider.cs

```diff
@@ -0,0 +1,333 @@
+        where TInvocationExpressionSyntax : SyntaxNode
+        where TObjectCreationExpressionSyntax : SyntaxNode
+    {
+        protected abstract ImmutableArray<string> TooManyArgumentsDiagnosticIds { get; }
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var diagnostic = context.Diagnostics.First();
+
+            var document = context.Document;
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            var initialNode = root.FindNode(diagnostic.Location.SourceSpan);
+
+            for (var node = initialNode; node != null; node = node.Parent)
+            {
+                if (node is TObjectCreationExpressionSyntax objectCreation)
+                {
+                    var argumentOpt = TryGetRelevantArgument(initialNode, node, diagnostic);
+                    await HandleObjectCreationExpressionAsync(context, objectCreation, argumentOpt).ConfigureAwait(false);
+                    return;
+                }
+                else if (node is TInvocationExpressionSyntax invocationExpression)
+                {
+                    var argumentOpt = TryGetRelevantArgument(initialNode, node, diagnostic);
+                    await HandleInvocationExpressionAsync(context, invocationExpression, argumentOpt).ConfigureAwait(false);
+                    return;
+                }
+            }
+        }
+
+        private TArgumentSyntax TryGetRelevantArgument(
+            SyntaxNode initialNode, SyntaxNode node, Diagnostic diagnostic)
+        {
+            if (this.TooManyArgumentsDiagnosticIds.Contains(diagnostic.Id))
+            {
+                return null;
+            }
+
+            return initialNode.GetAncestorsOrThis<TArgumentSyntax>()
+                              .LastOrDefault(a => a.AncestorsAndSelf().Contains(node));
+        }
+
+        private Task HandleInvocationExpressionAsync(
+            CodeFixContext context, TInvocationExpressionSyntax invocationExpression, TArgumentSyntax argumentOpt)
+        {
+            // Currently we only support this for 'new obj' calls.
+            return SpecializedTasks.EmptyTask;
+        }
+
+        private async Task HandleObjectCreationExpressionAsync(
+            CodeFixContext context,
+            TObjectCreationExpressionSyntax objectCreation,
+            TArgumentSyntax argumentOpt)
+        {
+            var document = context.Document;
+            var cancellationToken = context.CancellationToken;
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+
+            // Not supported if this is "new Foo { ... }" (as there are no parameters at all.
+            var typeNode = syntaxFacts.GetObjectCreationType(objectCreation);
```

Comment:
This doesn't seem like the right comment - (or it should be on that method).  There should be a type node in the example given...?